subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/EventFiringTaskExecuter.java

```diff
@@ -53,15 +56,25 @@ public TaskExecuterResult call(BuildOperationContext operationContext) {
             }
 
             private TaskExecuterResult executeTask(BuildOperationContext operationContext) {
+                Logger logger = task.getLogger();
+                ContextAwareTaskLogger contextAwareTaskLogger = null;
                 try {
                     taskExecutionListener.beforeExecute(task);
+                    BuildOperationRef currentOperation = buildOperationExecutor.getCurrentOperation();
+                    if (logger instanceof ContextAwareTaskLogger) {
+                        contextAwareTaskLogger = (ContextAwareTaskLogger) logger;
+                        contextAwareTaskLogger.setFallbackBuildOperationId(currentOperation.getId());
+                    }
                 } catch (Throwable t) {
                     state.setOutcome(new TaskExecutionException(task, t));
                     return TaskExecuterResult.WITHOUT_OUTPUTS;
                 }
 
                 TaskExecuterResult result = delegate.execute(task, state, context);
 
+                if (contextAwareTaskLogger != null) {
+                    contextAwareTaskLogger.setFallbackBuildOperationId(null);
```

Comment:
So if something is logged on this logger after the task is finished, it will fall back to the default build op? Is this the same as if something is logged before the task starts? Can we add some test coverage for these cases please?

Reply:
I've added test coverage. Its not strictly speaking the same as before the task. it depends on the `current` context